src/main/context/AuthContext.jsx

```diff
@@ -0,0 +1,42 @@
+import { createContext, useContext, useEffect, useState } from "react";
+
+import { GoogleAuthProvider, signInWithPopup, signOut, onAuthStateChanged } from "firebase/auth";
+import { auth } from "../firebase";
+
+const AuthContext = createContext()
+
+export const AuthContextProvider = ({ children }) => {
+  const [user, setUser] = useState()
+
+  // google signin
+  const googleSignIn = () => {
+    const provider = new GoogleAuthProvider();
+    signInWithPopup(auth, provider)
+  }
+
+  // signout
+  const logOut = () => {
+    signOut(auth)
+  }
+
+  // onAuthChange
+  useEffect(() => {
+    const unSubscribe = onAuthStateChanged(auth, (currentUser) => {
+      setUser(currentUser)
+    });
+    return () => {
+      unSubscribe()
+    };
+  },[])
+
+
+  return (
+    <AuthContext.Provider value={{googleSignIn,user,logOut}}>
+      {children}
+    </AuthContext.Provider>
+  )
+}
+
+export const UserAuth = () => {
```

Comment:
move this into src>hooks>useAuth.js

_this is called custom hooks_

Reply:
moved to hooks